src/NodeCompiler/CompileNodeToValue.php

```diff
@@ -36,63 +30,31 @@ class CompileNodeToValue
      */
     public function __invoke(Node $node, CompilerContext $context)
     {
-        if ($node instanceof Node\Scalar\String_
-            || $node instanceof Node\Scalar\DNumber
-            || $node instanceof Node\Scalar\LNumber) {
-            return $node->value;
-        }
-
-        // common edge case - negative numbers
-        if ($node instanceof Node\Expr\UnaryMinus) {
-            return $this($node->expr, $context) * -1;
-        }
-
-        if ($node instanceof Node\Expr\Array_) {
-            return $this->compileArray($node, $context);
-        }
-
-        if ($node instanceof Node\Expr\ConstFetch) {
-            return $this->compileConstFetch($node);
-        }
-
-        if ($node instanceof Node\Expr\ClassConstFetch) {
-            return $this->compileClassConstFetch($node, $context);
-        }
-
-        if ($node instanceof Node\Expr\BinaryOp) {
-            return $this->compileBinaryOperator($node, $context);
-        }
-
-        if ($node instanceof Node\Scalar\MagicConst\Dir) {
-            return $this->compileDirConstant($context);
+        if ($node instanceof Node\Stmt\Expression) {
+            return $this($node->expr, $context);
         }
 
-        if ($node instanceof Node\Scalar\MagicConst\Class_) {
-            return $this->compileClassConstant($context);
-        }
+        $constExprEvaluator = new ConstExprEvaluator(function (Node\Expr $node) use ($context) {
```

Comment:
Can the evaluator instance be preserved?

Reply:
No, because it needs the fallback evaluator with the right context :( I tried to implemented it without the fallback evaluator and call other evaluations in `catch` but it doesn't work. The test `CompileNodeToValueTest::testClassConstantResolutionWithAnotherClassConstant` was failing because the test is combination of multiplication (supported by PHP-Parser itself) and class constants (supported by BR).